File: Solutions - C/Plus One.c
/**
 * 0ms
Beats 100.00%
6.57MB
Beats 10.82%
 * Note: The returned array must be malloced, assume caller calls free().
 */

int* plusOne(int* digits, int digitsSize, int* returnSize){
    int i;
    int over = 1;

    for(i = digitsSize - 1; i >= 0; i--){
        if (over && digits[i] == 9){
            digits[i] = 0;
        }
        else if (over){
            digits[i] += 1;
            over = 0;
        }
    }

    if (!over){
        (*returnSize) = digitsSize;
        return (digits);
    }
    
    (*returnSize) = digitsSize + 1;
    int* res = (int *)malloc(sizeof(int) * (*returnSize));
    res[0] = 1;
    for (i = 1; i <  (*returnSize); i++){
        res[i] = 0;
    }

    return (res);
}

File: Solutions - Python/Plus One.py
"""
34ms
Beats 90.79%
16.21MB
Beats 59.93%
"""



class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        i = len(digits) - 1
        add = 1
        while (i >= 0):
            if (add and digits[i] == 9):
                digits[i] = 0
            elif(add):
                digits[i] += 1
                add = 0
                break
            i -= 1

        if (not add):
            return(digits)
        res = [0 for i in range(0, len(digits) + 1)]
        res[0] = 1
        return res